node/include/cocaine/detail/service/node/slave/machine.hpp

```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include <functional>
+#include <string>
+#include <system_error>
+
+#include <boost/circular_buffer.hpp>
+
+#include <asio/deadline_timer.hpp>
+#include <asio/io_service.hpp>
+#include <asio/posix/stream_descriptor.hpp>
+
+#include <cocaine/logging.hpp>
+
+#include "cocaine/idl/rpc.hpp"
+
+#include "cocaine/service/node/app/event.hpp"
+#include "cocaine/service/node/manifest.hpp"
+#include "cocaine/service/node/profile.hpp"
+#include "cocaine/service/node/slave/error.hpp"
+#include "cocaine/service/node/slave/id.hpp"
+
+#include "cocaine/detail/service/node/forwards.hpp"
+#include "cocaine/detail/service/node/splitter.hpp"
+
+namespace cocaine {
+namespace detail {
+namespace service {
+namespace node {
+namespace slave {
+
+using cocaine::service::node::slave::id_t;
+
+using detail::service::node::slave::channel_t;
+using detail::service::node::slave::control_t;
+using detail::service::node::slave::fetcher_t;
+using detail::service::node::slave::load_t;
+using detail::service::node::slave::state::state_t;
+
+/// Actual slave implementation.
+class machine_t : public std::enable_shared_from_this<machine_t> {
+    friend class detail::service::node::slave::state::active_t;
+    friend class detail::service::node::slave::state::handshake_t;
+    friend class detail::service::node::slave::state::seal_t;
+    friend class detail::service::node::slave::state::spawn_t;
+    friend class detail::service::node::slave::state::inactive_t;
+    friend class detail::service::node::slave::state::terminate_t;
+
+    friend class client_rpc_dispatch_t;
+
+    friend class detail::service::node::slave::channel_t;
+    friend class detail::service::node::slave::control_t;
+    friend class detail::service::node::slave::fetcher_t;
+
+public:
+    typedef std::function<void(std::uint64_t)> channel_handler;
+    typedef std::function<void(const std::error_code&)> cleanup_handler;
+
+private:
+    const std::unique_ptr<cocaine::logging::logger_t> log;
+
+public:
+    context_t& context;
+
+    // TODO: Drop namespaces.
+    const cocaine::service::node::slave::id_t id;
+    const profile_t profile;
+    const manifest_t manifest;
+
+private:
+    // TODO: In current implementation this can be invalid, when engine is stopped.
+    asio::io_service& loop;
+
+    /// The flag means that the overseer has been destroyed and we shouldn't call the callback.
+    std::atomic<bool> closed;
+    cleanup_handler cleanup;
+
+    splitter_t splitter;
+    synchronized<std::shared_ptr<fetcher_t>> fetcher;
+    boost::circular_buffer<std::string> lines;
+
+    std::atomic<bool> shutdowned;
```

Comment:
Personally I don't like atomics, but here is the right place not to touch ever no matter what.